Sources/tasks/CreatePost-Notify.php

```diff
@@ -152,71 +248,263 @@ public function execute()
 		{
-			$receiving_members[] = $done_member;
-			unset($watched[$done_member]);
+			$new_details = $this->_details;
+
+			if (empty($new_details['respawns']))
+				$new_details['respawns'] = 0;
+
+			if ($new_details['respawns']++ < 10)
+			{
+				$smcFunc['db_insert']('',
+					'{db_prefix}background_tasks',
+					array(
+						'task_file' => 'string',
+						'task_class' => 'string',
+						'task_data' => 'string',
+						'claimed_time' => 'int',
+					),
+					array(
+						'$sourcedir/tasks/CreatePost-Notify.php',
+						'CreatePost_Notify_Background',
+						$smcFunc['json_encode']($new_details),
+						max(0, $this->mention_mail_time - MAX_CLAIM_THRESHOLD),
+					),
+					array('id_task')
+				);
+			}
 		}
 
-		$parsed_message = array();
+		return true;
+	}
 
-		// Handle rest of the notifications for watched topics and boards
-		foreach ($watched as $member_id => $member_data)
+	private function updateAlerts($msg_id)
+	{
+		global $smcFunc;
+
+		// We send alerts only on the first iteration of this task.
+		if (!empty($this->_details['respawns']))
+			return;
+
+		// Delete alerts about any mentions and quotes that no longer exist.
+		if ($this->_details['type'] == 'edit')
+		{
+			$old_alerts = array();
+
+			$request = $smcFunc['db_query']('', '
+				SELECT content_action, id_member
+				FROM {db_prefix}user_alerts
+				WHERE content_id = {int:msg_id}
+					AND content_type = {literal:msg}
+					AND (content_action = {literal:quote} OR content_action = {literal:mention})',
+				array(
+					'msg_id' => $msg_id,
+				)
+			);
+			if ($smcFunc['db_num_rows']($request) != 0)
+			{
+				while ($row = $smcFunc['db_fetch_assoc']($request))
+					$old_alerts[$row['content_action']][$row['id_member']] = $row['id_member'];
+			}
+			$smcFunc['db_free_result']($request);
+
+			if (!empty($old_alerts))
+			{
+				$request = $smcFunc['db_query']('', '
+					SELECT content_type, id_mentioned
+					FROM {db_prefix}mentions
+					WHERE content_id = {int:msg_id}
+						AND (content_type = {literal:quote} OR content_type = {literal:msg})',
+					array(
+						'msg_id' => $msg_id,
+					)
+				);
+				while ($row = $smcFunc['db_fetch_assoc']($request))
+				{
+					$content_action = $row['content_type'] == 'quote' ? 'quote' : 'mention';
+					unset($old_alerts[$content_action][$row['id_mentioned']]);
+				}
+				$smcFunc['db_free_result']($request);
+
+				$smcFunc['db_query']('', '
+					DELETE FROM {db_prefix}user_alerts
+					WHERE content_id = {int:msg_id}
+						AND content_type = {literal:msg}
+						AND ((
+								content_action = {literal:quote}
+								AND id_member IN ({array_int:members_not_quoted})
+							)
+							OR (
+								content_action = {literal:mention}
+								AND id_member IN ({array_int:members_not_mentioned})
+							)
+						)',
+					array(
+						'msg_id' => $msg_id,
+						'members_not_quoted' => $old_alerts['quote'],
+						'members_not_mentioned' => $old_alerts['mention'],
+					)
+				);
```

Comment:
Looks like this will flail if either of these arrays is empty.

Reply:
Good catch.